sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala

```diff
@@ -142,9 +143,16 @@ case class QualifiedColType(
   def getV2Default: ColumnDefaultValue = {
     default.map { sql =>
       val e = ResolveDefaultColumns.analyze(colName, dataType, sql, "ALTER TABLE")
-      assert(e.resolved && e.foldable,
-        "The existence default value must be a simple SQL string that is resolved and foldable, " +
-          "but got: " + sql)
+      if (!e.resolved || !e.foldable) {
```

Comment:
This method is invoked in planning phase from DataSourceV2Strategy, and this was the first assertion on alter table path. I also modified second one (In CatalogV2Util)

Reply:
Hey so I looked at this code with Wenchen in detail.

Ultimately, we should not hit his assertion error ever. This happens at querying time, if any expression other than a literal value is stored in the column existence default metadata. This should never happen.

Instead, we should return a proper error message earlier here [1], when we try to run the ALTER TABLE ADD COLUMN command with a default value that we cannot constant-fold.

I can make a more involved fix later; for now, can you just update that location [1] to add a check if the result of the `ConstantFolding` is not a literal? It be something like:

```
      ConstantFolding(ReplaceExpressions(analyzed)) match {
        case e: Literal => e
        case _ =>
          throw new AnalysisException(
            errorClass = "COLUMN_DEFAULT_EXPRESSION_NOT_SUPPORTED",
           ...)
      }
```

This error class can have some message to indicate that the requested expression for the column default value is not supported by Spark, and the user should use a different expression there or remove the default entirely and then retry the command again.

[1] https://github.com/apache/spark/blob/5baaa615ef852c2032342c5986b4aa15ccf74b25/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala#L292

Reply:
Sure Daniel, sorry for late resonse

Reply:
I added isFoldable check to parsed expression. ConstantFolding is plan and its expression has alias which is not foldable, so I did check on parsed expression. Btw, do we need to check that it is resolved, since if something is foldable, that means no column reference are there, and I did not add that condition, is that ok?